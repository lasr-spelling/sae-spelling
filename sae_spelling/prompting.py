import random
from dataclasses import dataclass
from functools import partial
from typing import Callable


@dataclass
class SpellingPrompt:
    """
    Representation of a prompt used for spelling tasks. The prompt consists of a base, and answer, and a word.
    These fields might look like the following:

    base: "The word 'cat' is spelled:"
    answer: " c-a-t"
    word: "cat"

    The base may also contain ICL examples.
    """

    base: str
    answer: str
    word: str


def spelling(
    word: str,
    separator: str = "-",
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> str:
    """
    Break a word string into its component characters, separated by `char_separator`
    e.g. spelling("cat") -> " c-a-t"
    """
    if ignore_leading_space:
        word = word.strip()
    chars = list(word)
    if ignore_non_alpha_chars:
        chars = [c for c in chars if c.isalpha()]
    if capitalize:
        chars = [c.upper() for c in chars]
    return prefix + separator.join(chars)


def first_letter(
    word: str,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> str:
    """
    return just the first letter of the word, optionally capitalized
    e.g. first_letter("cat") -> " c"
    """
    if ignore_leading_space:
        word = word.strip()
    chars = list(word)
    if ignore_non_alpha_chars:
        chars = [c for c in chars if c.isalpha()]
    first_char = chars[0]
    if capitalize:
        first_char = first_char.upper()
    return prefix + first_char


def last_letter(
    word: str,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> str:
    """
    return just the last letter of the word, optionally capitalized
    e.g. last_letter("cat") -> " t"
    """
    if ignore_leading_space:
        word = word.strip()
    chars = list(word)
    if ignore_non_alpha_chars:
        chars = [c for c in chars if c.isalpha()]
    last_char = chars[-1]
    if capitalize:
        last_char = last_char.upper()
    return prefix + last_char


def is_present(
    word: str,
    char_to_check: str,
    prefix: str = " ",
    return_binary: bool = False,
) -> str:
    """
    Returns whether a character is present in the word or not
    e.g. is_present("cat", "t") -> "1"
    OR
    is_present("cat", "t") -> "True"
    """
    result = char_to_check in word

    return prefix + str(int(result)) if return_binary else prefix + str(result)


def letter_from_start(
    word: str,
    index: int,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> str:
    """
    return the letter of the word at the 'index' position relative to the START, optionally capitalized
    e.g. letter_from_start("mobile", 2) -> " b"
    """
    if ignore_leading_space:
        word = word.strip()

    chars = list(word)
    if ignore_non_alpha_chars:
        chars = [c for c in chars if c.isalpha()]

    char_at_idx = chars[index]

    if capitalize:
        char_at_idx = char_at_idx.upper()
    return prefix + char_at_idx


def letter_from_end(
    word: str,
    index: int,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> str:
    """
    return the letter of the word at the 'index' position relative to the END, optionally capitalized
    NOTE: This follows the Python notation of negative indexing
    i.e mobile[-1] will give 'e' and not 'l'
    e.g. letter_from_end("mobile", 2) -> " i"
    """
    if ignore_leading_space:
        word = word.strip()

    chars = list(word)
    if ignore_non_alpha_chars:
        chars = [c for c in chars if c.isalpha()]

    char_at_idx = chars[-index]

    if capitalize:
        char_at_idx = char_at_idx.upper()
    return prefix + char_at_idx


# ----- Formatters -------------------------------
Formatter = Callable[[str], str]


def spelling_formatter(
    separator: str = "-",
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> Formatter:
    return partial(
        spelling,
        separator=separator,
        prefix=prefix,
        capitalize=capitalize,
        ignore_leading_space=ignore_leading_space,
        ignore_non_alpha_chars=ignore_non_alpha_chars,
    )


def first_letter_formatter(
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> Formatter:
    return partial(
        first_letter,
        prefix=prefix,
        capitalize=capitalize,
        ignore_leading_space=ignore_leading_space,
        ignore_non_alpha_chars=ignore_non_alpha_chars,
    )


def last_letter_formatter(
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> Formatter:
    return partial(
        last_letter,
        prefix=prefix,
        capitalize=capitalize,
        ignore_leading_space=ignore_leading_space,
        ignore_non_alpha_chars=ignore_non_alpha_chars,
    )


def is_present_formatter(
    char_to_check: str,
    return_binary: bool = False,
    prefix: str = " ",
) -> Formatter:
    return partial(
        is_present,
        prefix=prefix,
        char_to_check=char_to_check,
        return_binary=return_binary,
    )


def letter_from_start_formatter(
    index,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> Formatter:
    return partial(
        letter_from_start,
        index=index,
        prefix=prefix,
        capitalize=capitalize,
        ignore_leading_space=ignore_leading_space,
        ignore_non_alpha_chars=ignore_non_alpha_chars,
    )


def letter_from_end_formatter(
    index,
    prefix: str = " ",
    capitalize: bool = False,
    ignore_leading_space: bool = True,
    ignore_non_alpha_chars: bool = True,
) -> Formatter:
    return partial(
        letter_from_end,
        index=index,
        prefix=prefix,
        capitalize=capitalize,
        ignore_leading_space=ignore_leading_space,
        ignore_non_alpha_chars=ignore_non_alpha_chars,
    )


# --------------------------------


def create_icl_prompt(
    word: str,
    examples: list[str],
    base_template: str = "{word}:",
    example_separator: str = "\n",
    answer_formatter: Formatter = spelling_formatter(),
    max_icl_examples: int | None = None,
    shuffle_examples: bool = True,
) -> SpellingPrompt:
    """
    Create a prompt with ICL examples in the base

    Args:
        word: the word to be spelled
        examples: a list of examples to use as ICL prompts. These will be shuffled
        base_template: a string template for the base of the prompt, including "{word}" as a placeholder for the word
        example_separator: a string to use to separate the ICL examples. default is newline
        answer_formatter: a function to format the answer. default is `spelling_formatter`, which spits out a string like " c-a-t" for the word "cat"
        max_icl_examples: the maximum number of ICL examples to use. If None, all examples will be used. default is None
        shuffle_examples: whether to shuffle the examples before selecting the first `max_icl_examples`. default is True
    """
    icl_prompts = []
    icl_examples = examples

    if max_icl_examples is not None:
        if shuffle_examples:
            icl_examples = random.sample(icl_examples, max_icl_examples)
        else:
            icl_examples = icl_examples[:max_icl_examples]
    elif shuffle_examples:
        icl_examples = examples.copy()
        random.shuffle(icl_examples)
    for ex in icl_examples:
        ex_answer = answer_formatter(ex)
        ex_base = base_template.format(word=ex)
        icl_prompts.append(ex_base + ex_answer)
    word_answer = answer_formatter(word)
    word_base = base_template.format(word=word)
    return SpellingPrompt(
        base=example_separator.join(icl_prompts) + example_separator + word_base,
        answer=word_answer,
        word=word,
    )


def random_icl_prompt(
    vocab: list[str],
    base_template: str = "{word}:",
    example_separator: str = "\n",
    answer_formatter: Formatter = spelling_formatter(),
    max_icl_examples: int = 10,
) -> SpellingPrompt:
    return create_icl_prompt(
        word=random.choice(vocab),
        examples=vocab,
        base_template=base_template,
        example_separator=example_separator,
        answer_formatter=answer_formatter,
        max_icl_examples=max_icl_examples,
    )
