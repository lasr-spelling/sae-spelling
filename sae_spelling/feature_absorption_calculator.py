import random
from dataclasses import dataclass
from typing import Callable

import torch
from sae_lens import SAE
from tqdm.autonotebook import tqdm
from transformer_lens import HookedTransformer

from sae_spelling.feature_attribution import (
    calculate_integrated_gradient_attribution_patching,
)
from sae_spelling.prompting import (
    Formatter,
    SpellingPrompt,
    create_icl_prompt,
    first_letter_formatter,
)
from sae_spelling.util import batchify

EPS = 1e-8


@dataclass
class FeatureScore:
    feature_id: int
    activation: float
    ablation_score: float
    probe_cos_sim: float


@dataclass
class AbsorptionResult:
    word: str
    prompt: str
    main_feature_scores: list[FeatureScore]
    top_ablation_feature_scores: list[FeatureScore]
    is_absorption: bool


@dataclass
class SampledAbsorptionResults:
    sample_portion: float
    main_feature_ids: list[int]
    sample_results: list[AbsorptionResult]


@dataclass
class FeatureAbsorptionCalculator:
    """
    Feature absorption calculator for spelling tasks.

    Absorption is defined by the following criteria:
    - The main features for a concept do not fire
    - There is a clear top feature by negative ablation score
    - The top feature is aligned with a probe trained on that concept
    """

    model: HookedTransformer
    icl_word_list: list[str]
    max_icl_examples: int | None = None
    base_template: str = "{word}:"
    answer_formatter: Formatter = first_letter_formatter()
    example_separator: str = "\n"
    shuffle_examples: bool = True
    # the position to read activations from (depends on the template)
    word_token_pos: int = -2
    filter_prompts_batch_size: int = 20
    ig_interpolation_steps: int = 6
    ig_batch_size: int = 6
    topk_feats: int = 10

    # the cosine similarity between the top ablation feature and the probe must be at least this high to count as absorption
    probe_cos_sim_threshold: float = 0.025
    # the top ablation score must be at least this much larger than the second highest score to count as absorption
    ablation_delta_threshold: float = 1.0

    @torch.inference_mode()
    def _filter_prompts(
        self,
        prompts: list[SpellingPrompt],
        sae: SAE,
        main_feature_ids: list[int],
    ) -> list[SpellingPrompt]:
        """
        Filter out any prompts where the main features are already active.
        NOTE: All prompts must have the same token length
        """
        self._validate_prompts_are_same_length(prompts)
        results: list[SpellingPrompt] = []
        for batch in batchify(prompts, batch_size=self.filter_prompts_batch_size):
            sae_in = self.model.run_with_cache([p.base for p in batch])[1][
                sae.cfg.hook_name
            ]
            sae_acts = sae.encode(sae_in)
            split_feats_active = (
                sae_acts[:, self.word_token_pos, main_feature_ids]
                .sum(dim=-1)
                .float()
                .tolist()
            )
            for prompt, res in zip(batch, split_feats_active):
                if res < EPS:
                    results.append(prompt)
        return results

    def _build_prompts(self, words: list[str]) -> list[SpellingPrompt]:
        return [
            create_icl_prompt(
                word,
                examples=self.icl_word_list,
                base_template=self.base_template,
                answer_formatter=self.answer_formatter,
                example_separator=self.example_separator,
                max_icl_examples=self.max_icl_examples,
                shuffle_examples=self.shuffle_examples,
            )
            for word in words
        ]

    def _ig_ablate(
        self,
        sae: SAE,
        prompt: SpellingPrompt,
        metric_fn: Callable[[torch.Tensor], torch.Tensor],
    ) -> tuple[torch.Tensor, torch.Tensor]:
        attrib = calculate_integrated_gradient_attribution_patching(
            model=self.model,
            input=prompt.base,
            include_saes={sae.cfg.hook_name: sae},
            metric_fn=metric_fn,
            patch_indices=self.word_token_pos,
            batch_size=self.ig_batch_size,
            interpolation_steps=self.ig_interpolation_steps,
        )
        with torch.no_grad():
            sae_acts = (
                attrib.sae_feature_activations[sae.cfg.hook_name][
                    0, self.word_token_pos
                ]
                .float()
                .cpu()
                .detach()
                .clone()
            )
            abl_scores = (
                attrib.sae_feature_attributions[sae.cfg.hook_name][
                    0, self.word_token_pos
                ]
                .float()
                .cpu()
                .detach()
                .clone()
            )
            return abl_scores, sae_acts

    def _is_absorption(
        self,
        main_feature_scores: list[FeatureScore],
        top_ablation_feature_scores: list[FeatureScore],
    ) -> bool:
        # if any of the main features fired, this isn't absorption
        if not all(score.activation < EPS for score in main_feature_scores):
            return False
        # if the top ablation score is positive, this isn't absorption
        # NOTE: should we change this? Should we only look at negative ablation scores to begin with?
        if top_ablation_feature_scores[0].ablation_score > 0:
            return False
        # If the top ablation score isn't significantly larger than the next largest score, this isn't absorption
        # NOTE: is this true? Is it possible to have 2 absorbing features firing at the same time?
        act_delta = abs(top_ablation_feature_scores[0].ablation_score) - abs(
            top_ablation_feature_scores[1].ablation_score
        )
        if act_delta < self.ablation_delta_threshold:
            return False
        # If the top firing feature isn't aligned with the probe, this isn't absorption
        if top_ablation_feature_scores[0].probe_cos_sim < self.probe_cos_sim_threshold:
            return False
        return True

    def calculate_absorption_sampled(
        self,
        sae: SAE,
        words: list[str],
        probe_dir: torch.Tensor,
        metric_fn: Callable[[torch.Tensor], torch.Tensor],
        main_feature_ids: list[int],
        max_ablation_samples: int | None = None,
        filter_prompts: bool = True,
        show_progress: bool = True,
    ) -> SampledAbsorptionResults:
        """
        This method calculates the absorption for each word in the list of words. If `max_ablation_samples` is provided,
        this method will randomly sample that many words to calculate absorption for as a performance optimization.
        If `filter_prompts` is True, this method will filter out any prompts where the main features are already active, as these cannot be absorption.
        """
        prompts = self._build_prompts(words)
        if filter_prompts:
            prompts = self._filter_prompts(prompts, sae, main_feature_ids)
        sample_portion = 1.0
        if max_ablation_samples is not None and len(prompts) > max_ablation_samples:
            sample_portion = max_ablation_samples / len(prompts)
            random.shuffle(prompts)
            prompts = prompts[:max_ablation_samples]
        results: list[AbsorptionResult] = []
        with torch.inference_mode():
            cos_sims = (
                torch.nn.functional.cosine_similarity(
                    probe_dir.to(sae.device), sae.W_dec, dim=-1
                )
                .float()
                .cpu()
            )
        for prompt in tqdm(prompts, disable=not show_progress):
            ig_scores, sae_acts = self._ig_ablate(sae, prompt, metric_fn=metric_fn)
            with torch.inference_mode():
                top_ig_feats = ig_scores.abs().topk(self.topk_feats).indices.tolist()
                main_feature_scores = _get_feature_scores(
                    main_feature_ids,
                    probe_cos_sims=cos_sims,
                    sae_acts=sae_acts,
                    ig_scores=ig_scores,
                )
                top_ablation_feature_scores = _get_feature_scores(
                    top_ig_feats,
                    probe_cos_sims=cos_sims,
                    sae_acts=sae_acts,
                    ig_scores=ig_scores,
                )
                is_absorption = self._is_absorption(
                    top_ablation_feature_scores=top_ablation_feature_scores,
                    main_feature_scores=main_feature_scores,
                )
                results.append(
                    AbsorptionResult(
                        word=prompt.word,
                        prompt=prompt.base,
                        main_feature_scores=main_feature_scores,
                        top_ablation_feature_scores=top_ablation_feature_scores,
                        is_absorption=is_absorption,
                    )
                )
        return SampledAbsorptionResults(
            sample_portion=sample_portion,
            main_feature_ids=main_feature_ids,
            sample_results=results,
        )

    def _validate_prompts_are_same_length(self, prompts: list[SpellingPrompt]):
        "Validate that all prompts have the same token length"
        token_lens = {len(self.model.to_tokens(p.base)[0]) for p in prompts}
        if len(token_lens) > 1:
            raise ValueError(
                "All prompts must have the same token length! Variable-length prompts are not yet supported."
            )


def _get_feature_scores(
    feature_ids: list[int],
    probe_cos_sims: torch.Tensor,
    sae_acts: torch.Tensor,
    ig_scores: torch.Tensor,
) -> list[FeatureScore]:
    return [
        FeatureScore(
            feature_id=feature_id,
            probe_cos_sim=probe_cos_sims[feature_id].item(),
            ablation_score=ig_scores[feature_id].item(),
            activation=sae_acts[feature_id].item(),
        )
        for feature_id in feature_ids
    ]
