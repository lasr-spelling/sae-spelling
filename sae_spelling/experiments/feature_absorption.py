from collections import defaultdict
from dataclasses import dataclass
from pathlib import Path

import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import torch
from sae_lens import SAE
from tqdm.autonotebook import tqdm
from transformers import PreTrainedTokenizerBase
from tueplots import axes, bundles

from sae_spelling.experiments.common import (
    EXPERIMENTS_DIR,
    PROBES_DIR,
    SaeInfo,
    get_gemmascope_saes_info,
    get_task_dir,
    humanify_sae_width,
    load_df_or_run,
    load_experiment_df,
    load_gemma2_model,
    load_gemmascope_sae,
    load_probe,
)
from sae_spelling.experiments.k_sparse_probing import (
    SPARSE_PROBING_EXPERIMENT_NAME,
    add_feature_splits_to_auroc_f1_df,
    get_sparse_probing_auroc_f1_results_filename,
    get_sparse_probing_raw_results_filename,
)
from sae_spelling.feature_absorption_calculator import FeatureAbsorptionCalculator
from sae_spelling.probing import LinearProbe
from sae_spelling.prompting import (
    VERBOSE_FIRST_LETTER_TEMPLATE,
    VERBOSE_FIRST_LETTER_TOKEN_POS,
    first_letter_formatter,
)
from sae_spelling.vocab import LETTERS, LETTERS_UPPER, get_alpha_tokens

FEATURE_ABSORPTION_EXPERIMENT_NAME = "feature_absorption"

# the cosine similarity between the top ablation feature and the probe must be at least this high
ABSORPTION_PROBE_COS_THRESHOLD = 0.025

# the top ablation score must be at least this much larger than the second highest score
ABSORPTION_FEATURE_DELTA_THRESHOLD = 1.0


@dataclass
class StatsAndLikelyFalseNegativeResults:
    probe_true_positives: int
    split_feats_true_positives: int
    split_feats: list[int]
    potential_false_negatives: list[str]


def letter_delta_metric(tokenizer: PreTrainedTokenizerBase, pos_letter: str):
    neg_letters = [
        f" {letter}" for letter in LETTERS_UPPER if pos_letter[-1].upper() != letter
    ]
    pos_letter_tok = tokenizer.encode(pos_letter)[-1]  # type: ignore
    neg_letter_toks = torch.tensor(
        [tokenizer.encode(neg_letter)[-1] for neg_letter in neg_letters]  # type: ignore
    )

    def metric_fn(logits):
        pos_logit = logits[:, -1, pos_letter_tok]
        neg_logits = logits[:, -1, neg_letter_toks]
        result = pos_logit - (neg_logits.sum(dim=-1) / len(neg_letters))
        return result

    return metric_fn


def calculate_ig_ablation_and_cos_sims(
    calculator: FeatureAbsorptionCalculator,
    sae: SAE,
    probe: LinearProbe,
    likely_negs: dict[str, StatsAndLikelyFalseNegativeResults],
    max_prompts_per_letter: int = 200,
) -> pd.DataFrame:
    results = []
    for letter, stats in tqdm(likely_negs.items()):
        assert calculator.model.tokenizer is not None
        absorption_results = calculator.calculate_absorption_sampled(
            sae,
            words=stats.potential_false_negatives,
            probe_dir=probe.weights[LETTERS.index(letter)],
            metric_fn=letter_delta_metric(calculator.model.tokenizer, letter),
            main_feature_ids=stats.split_feats,
            max_ablation_samples=max_prompts_per_letter,
            show_progress=False,
        )
        for sample in absorption_results.sample_results:
            top_feat_score = sample.top_ablation_feature_scores[0]
            second_feat_score = sample.top_ablation_feature_scores[1]
            result = {
                "letter": letter,
                "token": sample.word,
                "prompt": sample.prompt,
                "sample_portion": absorption_results.sample_portion,
                "num_probe_true_positives": stats.probe_true_positives,
                "split_feats": stats.split_feats,
                "split_feat_acts": [
                    score.activation for score in sample.main_feature_scores
                ],
                "split_feat_probe_cos": [
                    score.probe_cos_sim for score in sample.main_feature_scores
                ],
                "top_ablation_feat": top_feat_score.feature_id,
                "top_ablation_score": top_feat_score.ablation_score,
                "top_ablation_feat_probe_cos": top_feat_score.probe_cos_sim,
                "second_ablation_feat": second_feat_score.feature_id,
                "second_ablation_score": second_feat_score.ablation_score,
                "second_ablation_feat_probe_cos": second_feat_score.probe_cos_sim,
                "ablation_scores": [
                    score.ablation_score for score in sample.top_ablation_feature_scores
                ],
                "ablation_feats": [
                    score.feature_id for score in sample.top_ablation_feature_scores
                ],
                "ablation_feat_acts": [
                    score.activation for score in sample.top_ablation_feature_scores
                ],
                "ablation_feat_probe_cos": [
                    score.probe_cos_sim for score in sample.top_ablation_feature_scores
                ],
                "is_absorption": sample.is_absorption,
            }
            results.append(result)
    result_df = pd.DataFrame(results)
    return result_df


def get_stats_and_likely_false_negative_tokens(
    auroc_f1_df: pd.DataFrame,
    sae_info: SaeInfo,
    sparse_probing_task_output_dir: Path,
) -> dict[str, StatsAndLikelyFalseNegativeResults]:
    """
    Examine the k-sparse probing results and look for false-negative cases where the k top feats don't fire but our LR probe does
    """
    results: dict[str, StatsAndLikelyFalseNegativeResults] = {}
    raw_df = load_experiment_df(
        SPARSE_PROBING_EXPERIMENT_NAME,
        sparse_probing_task_output_dir
        / get_sparse_probing_raw_results_filename(sae_info),
    )
    for letter in LETTERS:
        split_feats = auroc_f1_df[auroc_f1_df["letter"] == letter]["split_feats"].iloc(  # type: ignore
            0
        )[0]
        k = len(split_feats)
        potential_false_negatives = raw_df[
            (raw_df["answer_letter"] == letter)
            & (raw_df[f"score_probe_{letter}"] > 0)
            & (raw_df[f"score_sparse_sae_{letter}_k_{k}"] <= 0)
        ]["token"].tolist()
        num_split_feats_true_positives = raw_df[
            (raw_df["answer_letter"] == letter)
            & (raw_df[f"score_probe_{letter}"] > 0)
            & (raw_df[f"score_sparse_sae_{letter}_k_{k}"] > 0)
        ].shape[0]
        num_probe_true_positives = raw_df[
            (raw_df["answer_letter"] == letter) & (raw_df[f"score_probe_{letter}"] > 0)
        ].shape[0]
        results[letter] = StatsAndLikelyFalseNegativeResults(
            probe_true_positives=num_probe_true_positives,
            split_feats_true_positives=num_split_feats_true_positives,
            split_feats=split_feats,
            potential_false_negatives=potential_false_negatives,
        )
    return results


def load_and_run_calculate_ig_ablation_and_cos_sims(
    calculator: FeatureAbsorptionCalculator,
    auroc_f1_df: pd.DataFrame,
    sae_info: SaeInfo,
    probes_dir: Path | str,
    sparse_probing_task_output_dir: Path,
) -> pd.DataFrame:
    layer = sae_info.layer
    probe = load_probe(layer=layer, probes_dir=probes_dir)
    sae = load_gemmascope_sae(layer, width=sae_info.width, l0=sae_info.l0)
    likely_negs = get_stats_and_likely_false_negative_tokens(
        auroc_f1_df,
        sae_info,
        sparse_probing_task_output_dir,
    )
    return calculate_ig_ablation_and_cos_sims(calculator, sae, probe, likely_negs)


def _aggregate_results_df(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
) -> pd.DataFrame:
    combined = []
    for layer_results in results.values():
        for df, sae_info in layer_results:
            if "sample_portion" not in df.columns:
                df["sample_portion"] = 1.0
            agg_df = (
                df[["letter", "is_absorption"]]
                .groupby(["letter"])
                .sum()
                .reset_index()
                .merge(
                    df[["letter", "sample_portion", "num_probe_true_positives"]]
                    .groupby(["letter"])
                    .mean()
                    .reset_index()
                )
            )
            agg_df["num_absorption"] = (
                agg_df["is_absorption"] / agg_df["sample_portion"]
            )
            agg_df["absorption_rate"] = (
                agg_df["num_absorption"] / agg_df["num_probe_true_positives"]
            )
            agg_df["layer"] = sae_info.layer
            agg_df["sae_width"] = sae_info.width
            agg_df["sae_l0"] = sae_info.l0
            combined.append(agg_df)
    df = pd.concat(combined)
    df["sae_width_str"] = df["sae_width"].map(humanify_sae_width)
    return df


def plot_absorption_rate_vs_l0(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
    experiment_dir: Path | str = EXPERIMENTS_DIR / FEATURE_ABSORPTION_EXPERIMENT_NAME,
    task: str = "first_letter",
):
    task_output_dir = get_task_dir(experiment_dir, task=task)
    df = _aggregate_results_df(results)

    sns.set_theme()
    plt.rcParams.update({"figure.dpi": 150})
    with plt.rc_context({**bundles.neurips2021(), **axes.lines()}):
        plt.figure(figsize=(3.75, 2.5))
        sns.scatterplot(
            df[["layer", "sae_l0", "sae_width_str", "absorption_rate"]]
            .groupby(["layer", "sae_width_str", "sae_l0"])
            .mean()
            .reset_index(),
            x="sae_l0",
            y="absorption_rate",
            hue="sae_width_str",
            s=15,
            rasterized=True,
        )
        plt.legend(title="SAE width", title_fontsize="small")
        plt.title("Mean absorption rate for first-letter task vs L0")
        plt.xlabel("L0")
        plt.ylabel("Mean absorption rate")
        plt.tight_layout()
        plt.savefig(task_output_dir / "absorption_rate_vs_l0.pdf")
        plt.show()


def plot_absorption_rate_vs_layer(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
    experiment_dir: Path | str = EXPERIMENTS_DIR / FEATURE_ABSORPTION_EXPERIMENT_NAME,
    task: str = "first_letter",
):
    task_output_dir = get_task_dir(experiment_dir, task=task)
    df = _aggregate_results_df(results)
    grouped_df = (
        df[["layer", "sae_l0", "sae_width_str", "absorption_rate"]]
        .groupby(["layer", "sae_width_str", "sae_l0"])
        .mean()
        .reset_index()
    )

    sns.set_theme()
    plt.rcParams.update({"figure.dpi": 150})
    with plt.rc_context({**bundles.neurips2021(), **axes.lines()}):
        plt.figure(figsize=(3.75, 2.5))
        sns.swarmplot(
            data=grouped_df,
            x="layer",
            y="absorption_rate",
            hue="sae_width_str",
            size=2,
        )

        # Customize the plot
        plt.title("Mean absorption rate for first-letter task vs Layer")
        plt.xlabel("Layer")
        plt.ylabel("Mean absorption rate")
        plt.legend(
            title="SAE width",
            bbox_to_anchor=(1.05, 1),
            loc="upper left",
            title_fontsize="small",
        )

        # Adjust layout to prevent clipping of the legend
        plt.tight_layout()
        plt.savefig(task_output_dir / "absorption_rate_vs_layer.pdf")
        plt.show()


def run_feature_absortion_experiments(
    layers: list[int],
    experiment_dir: Path | str = EXPERIMENTS_DIR / FEATURE_ABSORPTION_EXPERIMENT_NAME,
    sparse_probing_experiment_dir: Path | str = EXPERIMENTS_DIR
    / SPARSE_PROBING_EXPERIMENT_NAME,
    probes_dir: Path | str = PROBES_DIR,
    task: str = "first_letter",
    force: bool = False,
    skip_1m_saes: bool = True,
    skip_32k_saes: bool = True,
    skip_262k_saes: bool = True,
    skip_524k_saes: bool = True,
    feature_split_f1_jump_threshold: float = 0.03,
    verbose: bool = True,
) -> dict[int, list[tuple[pd.DataFrame, SaeInfo]]]:
    """
    NOTE: this experiments requires the results of the k-sparse probing experiments. Make sure to run them first.
    """
    task_output_dir = get_task_dir(experiment_dir, task=task)
    sparse_probing_task_output_dir = get_task_dir(
        sparse_probing_experiment_dir, task=task
    )

    model = load_gemma2_model()
    vocab = get_alpha_tokens(model.tokenizer)  # type: ignore
    calculator = FeatureAbsorptionCalculator(
        model=model,
        icl_word_list=vocab,
        max_icl_examples=10,
        base_template=VERBOSE_FIRST_LETTER_TEMPLATE,
        answer_formatter=first_letter_formatter(),
        word_token_pos=VERBOSE_FIRST_LETTER_TOKEN_POS,
        probe_cos_sim_threshold=ABSORPTION_PROBE_COS_THRESHOLD,
        ablation_delta_threshold=ABSORPTION_FEATURE_DELTA_THRESHOLD,
        ig_batch_size=6,
        ig_interpolation_steps=6,
        filter_prompts_batch_size=40,
    )
    results_by_layer: dict[int, list[tuple[pd.DataFrame, SaeInfo]]] = defaultdict(list)
    with tqdm(total=len(layers)) as pbar:
        for layer in layers:
            pbar.set_description(f"Layer {layer}")
            sae_infos = get_gemmascope_saes_info(layer)
            for sae_info in sae_infos:
                if skip_1m_saes and sae_info.width == 1_000_000:
                    continue
                if skip_32k_saes and sae_info.width == 32_000:
                    continue
                if skip_262k_saes and sae_info.width == 262_000:
                    continue
                if skip_524k_saes and sae_info.width == 524_000:
                    continue
                if verbose:
                    print(f"Running SAE {sae_info}", flush=True)
                auroc_f1_df = load_experiment_df(
                    SPARSE_PROBING_EXPERIMENT_NAME,
                    sparse_probing_task_output_dir
                    / get_sparse_probing_auroc_f1_results_filename(sae_info),
                )
                add_feature_splits_to_auroc_f1_df(
                    auroc_f1_df, feature_split_f1_jump_threshold
                )
                df_path = (
                    task_output_dir
                    / f"layer_{layer}_width_{sae_info.width}_l0_{sae_info.l0}.parquet"
                )
                df = load_df_or_run(
                    lambda: load_and_run_calculate_ig_ablation_and_cos_sims(
                        calculator,
                        auroc_f1_df,
                        sae_info,
                        probes_dir=probes_dir,
                        sparse_probing_task_output_dir=sparse_probing_task_output_dir,
                    ),
                    df_path,
                    force=force,
                )
                results_by_layer[layer].append((df, sae_info))
    return results_by_layer
