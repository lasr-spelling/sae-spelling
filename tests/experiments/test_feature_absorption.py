from sae_lens import SAE
from syrupy.assertion import SnapshotAssertion
from transformer_lens import HookedTransformer

from sae_spelling.experiments.feature_absorption import (
    calculate_ig_ablation_and_cos_sims,
)
from sae_spelling.feature_absorption_calculator import FeatureAbsorptionCalculator
from sae_spelling.probing import LinearProbe
from sae_spelling.prompting import (
    VERBOSE_FIRST_LETTER_TEMPLATE,
    VERBOSE_FIRST_LETTER_TOKEN_POS,
    first_letter_formatter,
)


def test_calculate_ig_ablation_and_cos_sims_gives_sane_results(
    gpt2_model: HookedTransformer, gpt2_l4_sae: SAE, snapshot: SnapshotAssertion
):
    fake_probe = LinearProbe(768, 26)
    calculator = FeatureAbsorptionCalculator(
        gpt2_model,
        icl_word_list=["dog", "cat", "fish", "bird"],
        base_template=VERBOSE_FIRST_LETTER_TEMPLATE,
        word_token_pos=VERBOSE_FIRST_LETTER_TOKEN_POS,
        answer_formatter=first_letter_formatter(),
    )
    # format: dict[letter: (num_true_positives, [split_feature_ids], [probable_feature_absorption_words])]
    likely_negs: dict[str, tuple[int, list[int], list[str]]] = {
        "a": (10, [1, 2, 3], [" Animal", " apple"]),
        "b": (100, [12], [" banana", " bear"]),
    }
    df = calculate_ig_ablation_and_cos_sims(
        calculator, gpt2_l4_sae, fake_probe, likely_negs
    )
    assert df.columns.values.tolist() == snapshot
