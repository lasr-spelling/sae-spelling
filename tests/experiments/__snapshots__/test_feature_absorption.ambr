# serializer version: 1
# name: test_calculate_ig_ablation_and_cos_sims_gives_sane_results
  list([
    'letter',
    'token',
    'prompt',
    'sample_portion',
    'num_true_positives',
    'split_feats',
    'split_feat_acts',
    'split_feat_probe_cos',
    'top_ablation_feat',
    'top_ablation_score',
    'top_ablation_feat_probe_cos',
    'second_ablation_feat',
    'second_ablation_score',
    'second_ablation_feat_probe_cos',
    'ablation_scores',
    'ablation_feats',
    'ablation_feat_acts',
    'ablation_feat_probe_cos',
    'is_absorption',
  ])
# ---
